### PR TITLE
move vision code to nupic.vision repo

### LINKED ISSUE #1863: Move vision code to appropriate repo nupic.vision
and remove it from nupic. 
Relevant task: https://github.com/nupic-community/nupic.vision/issues/3

Possibly depends on https://github.com/numenta/nupic/issues/1978

### PR DESCRIPTION
**Pull request by breznak**
and remove from nupic

Fixes #1863 
Blocked by: https://github.com/nupic-community/nupic.vision/issues/8

### CODE REVIEW

**Comment by rhyolight** [comment]
I believe this PR would be better merged after https://github.com/numenta/nupic/issues/1978 is complete, from discussion of this issue at #1863.

**Comment by david-ragazzi** [comment]
> I believe this PR would be better merged after #1978 is complete, from discussion of this issue at #1863.

@rhyolight With https://github.com/numenta/nupic.core/pull/438 and https://github.com/numenta/nupic/pull/2165 merged, could @breznak continue his (nice) work moving vision code (https://github.com/numenta/nupic/pull/1866)?

**Comment by rhyolight** [comment]
@david-ragazzi I'd say yes as long as #1978 is obsolete (I've asked that question just to make sure).

**Comment by rhyolight** [comment]
@breznak This PR is ready to move again. Can you please look it over and decide if it is still ready for review or if you need to update anything first? Thanks!

**Comment by breznak** [comment]
Thanks Matt! I'll review and ping when ready (but this has lower pr. for me now)

File: examples/opf/experiments/multistep/datasets/first_order_0.csv [hunk 1]
```diff
@@ -2,14 +2,11 @@ reset,field1,field2
 int,string,float
 R,,
 1,0,0
-0,1,1
-0,2,2
-0,0,0
```
Comment (rhyolight) [inline examples/opf/experiments/multistep/datasets/first_order_0.csv]: Why all the changes to this file? I'm not certain this was meant to be included in this PR.

**Comment by rhyolight** [comment]
@breznak With this removed, can we also remove [`pillow` from `requirements.txt`](https://github.com/numenta/nupic/blob/master/external/common/requirements.txt#L6)?

**Comment by rhyolight** [comment]
Once this is merged, we can close out the super issue https://github.com/numenta/nupic/issues/1862!

**Comment by breznak** [comment]
> @breznak With this removed, can we also remove pillow from requirements.txt?

yop, just did that. When we get to merging this PR, do you think we should make a tag "vision_removed" so it's easier to check back if we needed?

**Comment by breznak** [comment]
blocked by https://github.com/nupic-community/nupic.vision/issues/8

**Comment by scottpurdy** [comment]
How about we mark it as deprecated now and we can delete the files in a few weeks to give people a chance to switch to nupic.vision after we clean it up.

**Comment by jaredweiss** [comment]
Looks good to me. [numenta/nupic.vision#8](https://github.com/numenta/nupic.vision/issues/8) was solved when [numenta/nupic.core#359](https://github.com/numenta/nupic.core/issues/359) was fixed. There's a good example [here](https://github.com/numenta/nupic/blob/8bd9000d4c95ff5af19224a9542d98afc71607df/examples/network/custom_region_demo.py) and my PR in nupic.vision also has the image sensor region working.

**Comment by scottpurdy** [comment]
We need to have a period where nupic.vision is stable and works instead of the version in nupic. The versioning of NuPIC hopefully protects us but a couple weeks delay is much safer and has no real downside.

Can we close this for now?

**Comment by breznak** [comment]
@scottpurdy closing as you suggested. Someone should add a not to the deprecated files/print a debug msg when initialized

**Comment by scottpurdy** [comment]
@breznak - I think it has been long enough that we can remove the vision stuff now. Feel free to send a PR.

**Comment by breznak** [comment]
@scottpurdy does it work like this?

**Comment by scottpurdy** [comment]
Looks good but PCANode was not moved to nupic.vision as far as I see.

**Comment by breznak** [comment]
> Looks good but PCANode was not moved to nupic.vision as far as I see.

@scottpurdy thanks for spotting that! Giving it a closer look, it is likely unused now, but is not purely vision-related. So `PCANode` should rather stay here, correct?

File: nupic/regions/PCANode.py [hunk 2]
```diff
@@ -435,36 +435,3 @@ def getOutputElementCount(self, name):
...
-
-  from nupic.engine import *
-
-  rgen = numpy.random.RandomState(37)
-
-  inputSize = 8
-
-  net = Network()
-  sensor = net.addRegion('sensor', 'py.ImageSensor' ,
-          '{ width: %d, height: %d }' % (inputSize, inputSize))
```
Comment (breznak) [inline nupic/regions/PCANode.py]: @scottpurdy ok, disregard my last comment. This PR does keep `PCANode` here where it should stay. It is just removing (already broken) example from "main". Or do you want me to just comment this out to keep the code here?

Comment (scottpurdy) [inline nupic/regions/PCANode.py]: I agree that we should delete the main. Should we turn it into a test? Or put a simplified example in the module doc string?

Comment (breznak) [inline nupic/regions/PCANode.py]: added as a test from PCANode in nupic, without the `ImageSensor` stuff which is from nupic.vision; 
in the last commit, but I haven't tested it at all, as I dont have a running nupic now :sob:

File: tests/unit/nupic/regions/pcanode_test.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import unittest2 as unittest
+
+from nupic.regions.PyRegion import PyRegion
+from nupic.engine import *
+
+class PCANodeTest(unittest.TestCase):
+
+
+ def testPCANode(self):
+  rgen = numpy.random.RandomState(37)
```
Comment (scottpurdy) [inline tests/unit/nupic/regions/pcanode_test.py]: this isn't used, remove it

File: tests/unit/nupic/regions/pcanode_test.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+import unittest2 as unittest
+
+from nupic.regions.PyRegion import PyRegion
+from nupic.engine import *
+
+class PCANodeTest(unittest.TestCase):
+
+
+ def testPCANode(self):
```
Comment (scottpurdy) [inline tests/unit/nupic/regions/pcanode_test.py]: Use 2 space indents

File: tests/unit/nupic/regions/pcanode_test.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import unittest2 as unittest
```
Comment (scottpurdy) [inline tests/unit/nupic/regions/pcanode_test.py]: `import unittest`

File: tests/unit/nupic/regions/pcanode_test.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2014, Numenta, Inc.  Unless you have an agreement
```
Comment (scottpurdy) [inline tests/unit/nupic/regions/pcanode_test.py]: 2015

File: tests/unit/nupic/regions/pcanode_test.py (hunk 3 above)
Comment (scottpurdy) [inline tests/unit/nupic/regions/pcanode_test.py]: oh nvm it is needed to seed the RNG, just don't need to store the return value in `rgen`

File: tests/unit/nupic/regions/pcanode_test.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  inputSize = 8
+
+  net = Network()
+  sensor = net.addRegion('sensor', 'py.ImageSensor' ,
+          '{ width: %d, height: %d }' % (inputSize, inputSize))
+
+  params = """{bottomUpCount: %d,
+              SVDSampleCount: 5,
+              SVDDimCount: 2}""" % inputSize
```
Comment (scottpurdy) [inline tests/unit/nupic/regions/pcanode_test.py]: I would think the bottomUpCount would be `inputSize * inputSize`, no?

Comment (scottpurdy) [inline tests/unit/nupic/regions/pcanode_test.py]: I suppose the SVDDimCount might account for that... I guess if it works then it's fine as is

File: tests/unit/nupic/regions/pcanode_test.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+  linkParams = '{ mapping: in, rfSize: [%d, %d] }' % (inputSize, inputSize)
+  net.link('sensor', 'pca', 'UniformLink', linkParams, 'dataOut', 'bottomUpIn')
+
+  net.initialize()
+
+  for i in range(10):
+    pca.getSelf()._testInputs = numpy.random.random([inputSize])
+    net.run(1)
+    #print s.sendRequest('nodeOPrint pca_node')
```
Comment (scottpurdy) [inline tests/unit/nupic/regions/pcanode_test.py]: remove?

File: tests/unit/nupic/regions/pcanode_test.py (hunk 7 above)
Comment (breznak) [inline tests/unit/nupic/regions/pcanode_test.py]: yep, I have no idea.

**Comment by breznak** [comment]
@scottpurdy sorry, I forgot about this PR. Merged and updated.

**Comment by scottpurdy** [comment]
Waiting to hear whether @breznak thinks PCANode and test should be in nupic or nupic.vision

**Comment by rhyolight** [comment]
@scottpurdy Thanks for the timely update.

**Comment by breznak** [comment]
> Waiting to hear whether @breznak <https://github.com/breznak> thinks
PCANode and test should be in nupic or nupic.vision

I believe it's not used in nupic, so I'm for the move to nupic.vision. I
won't have time to do it myself over the weekend though

**Comment by breznak** [comment]
@rhyolight I've tried moving the test to `nupic.vision` - does not help either. There is still some bug (which I won't have time to debug :() .

```
ERR:  Network::initialize() -- unable to evaluate all links                                                                                                                                                                      
The following links could not be evaluated:                                                                                                                                                                                      
 [/home/travis/build/numenta/nupic.core/src/nupic/engine/Network.cpp line 514]                                                                                                                                                   
Traceback (most recent call last):                                                                                                                                                                                               
  File "nupic/vision/regions/PCANode.py", line 465, in <module>                                                                                                                                                                  
    net.initialize()                                                                                                                                                                                                             
  File "/home/mmm/devel/nupic/nupic-source/nupic/engine/__init__.py", line 692, in initialize                                                                                                                                    
    engine.Network.initialize(self, *args, **kwargs)
  File "/home/mmm/devel/nupic/env/lib/python2.7/site-packages/nupic.bindings-0.1.5-py2.7-linux-x86_64.egg/nupic/bindings/engine_internal.py", line 1078, in initialize
    return _engine_internal.Network_initialize(self)
RuntimeError: Network::initialize() -- unable to evaluate all links
The following links could not be evaluated:
 (the list is actually empty)
```

I'm suggesting: 
a) delete the (not working) test and keep PCANode in nupic.
b) remove PCANode class as well, it's nowhere used in code. (imho better)

**Comment by rhyolight** [comment]
My opinion is that we already have a pretty packed codebase, so a more liberal cleanup is better. We can always add it back of people need it later. So I vote for (b), but I must defer to @numenta/nupic-reviewers .

**Comment by subutai** [comment]
> I'm suggesting: 
> a) delete the (not working) test and keep PCANode in nupic.
> b) remove PCANode class as well, it's nowhere used in code. (imho better)

PCANode is not specific to vision but it is not currently used anywhere either. I'm not sure when we would use it again so I'm fine with (b) as well. Should it be tracked as a separate issue?

**Comment by rhyolight** [comment]
> Should it be tracked as a separate issue?

I think @breznak can just update this PR, right?

**Comment by scottpurdy** [comment]
I have separate PRs to handle PCANode:

- #2560 
- numenta/nupic.vision#42

**Comment by breznak** [comment]
Thanks Scott for taking care of the issues. 
@rhyolight I resolved the big recent merge, should be ready.

**Comment by rhyolight** [comment]
@scottpurdy Can you give this a thumbs up now?

**Comment by rhyolight** [comment]
I :heart: closing a 6-month-old PR :+1:

**Pull request merged**